PostCreatingPage 생성

NJ97S opened the pull request:
## 🚀 작업 내용

- [x] PostCreatingPage 데스크탑 버전으로 구현 완료하였습니다.

## 📝 참고 사항

- 아직 코드 수정이 더 필요할 것 같은데, 큰 틀은 만들어져서 리팩토링 관련 의견 받고자 PR 드립니다 🥹

## 🖼️ 스크린샷

![RollingPaper-Chrome2024-02-2921-57-04-ezgif com-video-to-gif-converter](https://github.com/CreativePaperCrew/RollingPaper/assets/79499733/e0e40003-00b1-4086-a31b-6b305e531e3f)


## 🚨 관련 이슈

- 최초 렌더링 시에도 내용을 입력해달라는 에러 메세지가 출력되고 있습니다... 수정하겠습니다..
- 아직 api 연동은 되어있지 않습니다.

## ✅ 이후 계획

- ~~`header`와 `button` component가 merge 되는 대로 적용하여 페이지 완성할 예정입니다.~~ -> 완료
- 태블릿과 모바일 환경에 맞춰 모바일 쿼리 작업 진행할 예정입니다.

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
feat: `/post` 페이지 틀 생성 (기능 구현 X)

NJ97S committed:
feat: color box 선택 기능 추가

- color box 클릭하면 해당 box에 체크 아이콘 표시

NJ97S committed:
feat: 컬러 또는 이미지 선택 기능 추가

- 컬러 또는 이미지 버튼 클릭하였을 때, 선택 효과 표시
- 아직 기능 구현까지는 되지 않았음.

NJ97S committed:
feat: 컬러 버튼 눌렀을 때, color box 보여주는 기능 추가

NJ97S committed:
feat: 배경화면 이미지 선택 기능 추가

NJ97S committed:
refactor: BackfounrdSelector.jsx state 합침

- selected state의 경우, 하나의 state로 관리가 가능하여 통합하였음.

NJ97S committed:
refactor: 불필요한 렌더링을 피하기 위해 useCallback 추가

- PostCreatingPage.jsx 내 함수의 불필요한 재렌더링을 방지하기 위해,
- useCallback을 사용하였음.

vinoankr commented on src/components/ToggleButton/ToggleButton.jsx:
```diff
@@ -0,0 +1,42 @@
+import * as S from './ToggleButtonStyle';
+
+const ToggleButton = ({
```
ToggleButton은 네이밍이 너무 general한 것 같아요. 토글 버튼은 코드 다른 곳들에도 무수히 쓰이니깐 무슨 일을 하는지 명확히 명명해주면 좋을 것 같아요.

vinoankr commented on src/pages/PostCreatingPage/PostCreatingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+            isColorSelected={isColorSelected}
+            isImageSelected={isImageSelected}
+          />
+          {
```
중괄호가 큰 의미 없는 것 같아요

vinoankr commented on src/pages/PostCreatingPage/PostCreatingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+import ColorBox from '../../components/BackgroundSelector/BackgroundSelector';
+
+const PostCreatingPage = () => {
+  const [isColorSelected, setIsColorSelected] = useState(false);
+  const [isImageSelected, setIsImageSelected] = useState(false);
```
useState 두 가지를 사용하는 것보다 한 개로 통합해서 값을 'color' / 'image' / null 로 설정하면 좋을 것 같아요

vinoankr commented on src/pages/PostCreatingPage/PostCreatingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+            </S.InformationMessage>
+          </S.ContainerSelect>
+          <ToggleButton
+            handleColorClick={handleColorClick}
+            handleImageClick={handleImageClick}
+            isColorSelected={isColorSelected}
+            isImageSelected={isImageSelected}
```
useState을 간결하게 수정하면 여기서 넘어가는 프롭은 한 개면 충분할 것 같아요

NJ97S replied:
최대한 넘겨주는 prop을 하나로 줄여보고자 하였으나, `handleColorClick`과 `handleImageClick`을 하나로 통합하지 못했습니다ㅠㅠ 수정된 코드 한 번 더 확인해주시면 감사하겠습니다 🥹

vinoankr commented on src/pages/PostCreatingPage/PostCreatingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+    <>
+      <header></header>
+      <main>
+        <S.ContainerWhole>
```
Container라는 단어를 스타일드 컴포넌트에 중첩해서 사용하시는데, 적합하지 않아보입니다. container는 감싸는 요소에서 한 번만 사용되면 좋을 것 같아요

vinoankr commented on src/pages/PostCreatingPage/PostCreatingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+            isImageSelected={isImageSelected}
+          />
+          {
+            <ColorBox
```
BackgroundSelector의 오타인가요?

NJ97S replied:
넵... 수정하다가 놓친 것 같습니다... 죄송함다... 😭

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
feat: PostCreatingPage 생성하기 버튼 추가

- 아직 링크 연결 안되어 있음.

NJ97S committed:
refactor: ToggleButton component 이름 수정

- BackgroundTypeSelectButton으로 이름 수정

NJ97S committed:
refactor: 불필요한 코드 삭제

- PostCreatingPage.jsx 내의 dead code 제거

NJ97S committed:
fix: color와 image 값이 같이 넘어가는 이슈 해결

- color를 선택하였을 때, image 기본값이 같이 넘어가는 이슈가 있어,
- useEffect 함수 내 조건문 추가 및 의존성 배열 수정을 통해 해결하였음.

NJ97S committed:
refactor: dead code 제거

- BackgroundSelector.jsx 파일 내 console 코드 삭제

NJ97S committed:
feat: 배경 이미지 선택 시 선택 효과 추가

- PostCreatingPage에서 배경 이미지 선택 시, opacity 0.5 적용

Trophy198 commented on src/components/BackgroundSelector/BackgroundSelector.jsx:
```diff
@@ -0,0 +1,65 @@
+  { name: 'purple', value: '--purple-200' },
+  { name: 'blue', value: '--blue-200' },
+  { name: 'green', value: '--green-200' },
+];
```
해당하는 상수부분이 생각보다 쓰이는곳이 많은데 이를 상수로 분리하고 객체가 변조되지 않게 동결시키면 어떨까요?

Trophy198 commented on src/components/BackgroundSelector/BackgroundSelector.jsx:
```diff
@@ -0,0 +1,65 @@
+  return (
+    <S.BoxContainer>
+      {console.log(selectedColor)}
+      {console.log(selectedImage)}
```
테스트를 완료한 이후엔 `console.log`를 지워주세요!🧐

NJ97S replied:
넵ㅠㅠㅠ 지우겠습니다ㅠㅠ

Trophy198 commented on package.json:
```diff
@@ -7,7 +7,6 @@
     "@testing-library/react": "^13.4.0",
     "@testing-library/user-event": "^13.5.0",
     "axios": "^1.6.7",
-    "emoji-picker-react": "^4.8.0",
```
`emoji-picker-react`는 사용하고 있는 라이브러리인데 여기선 삭제가 된것 같아요......👀

Trophy198 reviewed (commented):
수고많으셨습니다! 나머지 이야기할 부분은 추후 미팅을 통해 얘기드리도록 하겠습니다.
아직 수정해야할 부분이 있으니 그부분들은 수정 부탁드릴게요!
추가적으로 `BackgroundSelector.jsx` 와 같은 컴포넌트에서 
```javascript
 useEffect(() => {
    apiClient.get('/background-images/').then((res) => {
      setImageUrls(res.data.imageUrls);
      if (select === 'image') {
        setSeletedImage(res.data.imageUrls[0]);
      } else {
        setSeletedImage(null);
      }
    });
  }, [select]);
 ```
 해당 api호출 부분을 사용하고 있는것을 보았습니다. 
 제가 생각하기에 `BackgroundSelector`가 어떤역할인지 생각해보면 백그라운드 색상과 이미지를 선택하는데 사용이되는거 같은데 현재 코드는 외부 api를 호출하고 해당 api의 응답을 기반으로 이미지URL을 설정하고 초기화하는 부분이 있는것 같습니다.
 
 재가 생각하기엔 외부 api 호출 부분을 분리하여 해당 컴포넌트는 부모 컴포넌트에게 이미지URL을 전달하도록 하게끔하면 이미지 URL을 직접 관리하지 않아도 부모 컴포넌트에서 받아와 좀더 컴포넌트? 같은 모습이 되는것 같아요! 
 
 위 부분 한번 생각해보시면 좋을것 같아요!

NJ97S committed:
refactor: COLORS 상수 파일로 분리

- BackgroundSelector.jsx 내의 COLORS 상수를 파일로 따로 분리함.

NJ97S committed:
refactor: styled component 이름 수정

NJ97S committed:
refactor: api 분리 및 에러 메세지 처리

NJ97S committed:
refactor: useEffect 함수 분리

- PostCreatingPage.jsx 내 useEffect 함수를 분리하였습니다.

NJ97S committed:
refactor: COLORS 객체 변경에 따른 코드 리팩토링

- COLORS 상수화 및 객체 동결
- COLORS 객체 형태에 맞춰, Object.entries 사용하여 코드 리팩토링 진행함.

JaeBeen95 commented on src/pages/PostCreatingPage/PostCreatingPageStyle.js:
```diff
@@ -0,0 +1,37 @@
+  margin-top: 3.5625rem;
+`;
+
+export const Receiver = styled.div`
```
제안: api에서 recipeint라는 단어를 사용하고 있으니 해당 네이밍을 바꿔보면 어떨까요?

NJ97S committed:
Merge branch 'main' of https://github.com/CreativePaperCrew/RollingPaper into snj/PostCreatingPage

NJ97S committed:
feat: input에 값 없을 경우, 에러 효과 추가

- input에 값 없을 경우, 에러 메세지 출력 및 버튼 비활성화

yeona813 reviewed (approved):
수고하셨습니다!  버튼 컴포넌트 disabled 추가를 미처 못한 점 추가해주셔서 감사해요🥰🥰 코드에서 제가 참고할 부분이 정말 많은 것 같습니다 👍🏻👍🏻👍🏻

NJ97S committed:
Merge branch 'main' into snj/PostCreatingPage
